### PR TITLE
[CR] Updated StatsFromSkills to use set_value & get_value for base stats

### PR DESCRIPTION
**Pull request by macrosblackd**
This updates the StatsFromSkills mod to utilize the set_value & get_value methods to store the player's base stat values.

I'm in need of suggestions for handling converting old characters using statsfromskills (which directly modified *_max) in a way that won't screw with new characters.

### CODE REVIEW

**Comment by chaosvolt** [comment]
Hmm. So this essentially means character creation is no longer borked with it? That would eliminate half of what makes the mod so OP, yeah.

**Comment by Coolthulhu** [comment]
> I'm in need of suggestions for handling converting old characters using statsfromskills (which directly modified *_max) in a way that won't screw with new characters.

Add a new lua callback on game start, then bind a command to it. Just add `lua_callback("on_new_game");` at the bottom of `game::start_game`.
While we're at it, adding a lua callback on skill gain could also be helpful for this mod.

**Comment by macrosblackd** [comment]
Something else I came across - it seems mutations are directly edititing stat_max values as well - which means I would need to implement them in the lua or come up with some other way of either handling the stats from skills or change the way mutations adjust stats.

**Comment by Coolthulhu** [comment]
You could calculate the bonus from skills and save it separately, then at next recalculation you could subtract it to get base stats+mut stats.

**Comment by macrosblackd** [comment]
Is there a way to reapply mutations?

**Comment by Coolthulhu** [comment]
Nope, stat mutations apply their effect when added and apply the inverse of it when removed.
You have to keep the stat-from-skill bonus separate.

**Comment by macrosblackd** [comment]
That doesn't leave me an easy way to reset their base stats to 8 if they used the previous version :(

**Comment by Coolthulhu** [comment]
You can just un-apply the skill bonus, save what's left as base stats, then re-apply the skill bonus. That is, assuming you are calculating it the same way.
That may require editing the bonus applying function to return the bonus first rather than applying it immediately, then applying the bonus outside the calculating function.

And for the future, it may be better to save `player:set_value("StatsThoughSkills", [version here])` instead of `true`. That way if something else changes, it will be easier to detect from which version did it change and what updates does it require.

**Comment by macrosblackd** [comment]
I'm having problems with a critical lua error that I can't seem to track down - any chance I could get a hand from someone?

**Comment by Coolthulhu** [comment]
Post the error

**Comment by macrosblackd** [comment]
It's not giving me an error, it's just crashing. I've tried with the debugger too - still nothing.

File: .gitignore [hunk 1]
```diff
@@ -104,3 +104,6 @@ cataclysm.a
 
 # coredumps
 *.core
+
+# sublime-text workspace
+*.sublime-workspace
```
Comment (Malkeus) [inline .gitignore]: I don't think you'll want to add the sublime workspaces for the whole project to .gitignore, they will probably need to be updated in the future, I know the codeblocks project file gets modified frequently. If you had to update the workspace to make the project work, include those changes for review. 
Iirc several libraries were added to the codeblocks project not to long ago, I don't think the sublime workspace was updated at the same time.

Comment (macrosblackd) [inline .gitignore]: The workspaces is the user specific settings.

Comment (Malkeus) [inline .gitignore]: Oh, ok then. I don't know anything about sublime. Thanks for the info :)

**Comment by Malkeus** [comment]
I merged this, I'm getting a crash when I gain a skill with a character that started with the original version. I'll attempt a debug build, but I could use some advice on using codeblocks to trace something like this.

**Comment by Malkeus** [comment]
I'm stuck, can't operate the debugger in codeblocks. I'll have to do more research tomorrow. This is what I get from windows when the game crashes, it's in ntdll.dll.
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	Cataclysm-Tiles.exe
  Application Version:	0.0.0.0
  Application Timestamp:	56a49b77
  Fault Module Name:	ntdll.dll
  Fault Module Version:	6.1.7601.18798
  Fault Module Timestamp:	5507b864
  Exception Code:	c0000028
  Exception Offset:	00000000000c9018
  OS Version:	6.1.7601.2.1.0.256.1
  Locale ID:	1033
  Additional Information 1:	523e
  Additional Information 2:	523eb175b1da9e38c993564e88987f1d
  Additional Information 3:	5aaa
  Additional Information 4:	5aaa4b51560ab6743aca25b616736e31

**Comment by macrosblackd** [comment]
This seems to be the same crash I was getting, I thought I fixed it since I
wasn't getting it though. I'm going to keep trying on Linux to reproduce
since that's the only place I can see the errors.

**Comment by macrosblackd** [comment]
Found the issue and fixed! I also added a skill gain callback when reading.

**Comment by Malkeus** [comment]
Awesome! I'll merge this later this evening and try it out. I should have
mentioned my operating system, I'm on windows 7, 64 bit.

File: lua/class_definitions.lua [hunk 2]
```diff
@@ -75,6 +75,24 @@ classes = {
...
+        attributes = {
+        },
+        functions = {
+            { name = "get_turn", rval = "int", args = { } },
+            { name = "minutes_past_midnight", rval = "int", args = { } },
+            { name = "seconds_past_midnight", rval = "int", args = { } },
+            { name = "sunrise", rval = "calendar", args = { } },
+            { name = "sunset", rval = "calendar", args = { } },
+            { name = "is_night", rval = "bool", args = { } },
+            { name = "sunlight", rval = "calendar", args = { } },
```
Comment (BevapDin) [inline lua/class_definitions.lua]: Those functions here return a `calendar` object by value, but the Lua wrapper stores them all as references (se header of lua/class_definitions.lua). When the functions get called from Lua, the wrapper stores a reference to a temporary object in Lua, which is immediately destroy, leaving an invalid reference behind.


`get_calendar_turn_wrapper` is fine, it returns a reference to the global calendar object.

Comment (macrosblackd) [inline lua/class_definitions.lua]: is there a way to flag those functions as byref?

Comment (BevapDin) [inline lua/class_definitions.lua]: The by-ref vs by-value thing is (currently) part of the type. If type X is defined as by-value, all usage of that type will be by-value. A per-function setting is not supported (feel free to change this).

Comment (macrosblackd) [inline lua/class_definitions.lua]: How would I go about adding this? I was looking through the generate_binding code, but I'm not entirely sure what needs to be done.

Comment (BevapDin) [inline lua/class_definitions.lua]: From the top of my head, you need to
- add some markers in lua/classbindings.lua to distinguish function that return by value and those that return by reference.
- add code in generate_binding.lua to generate the proper wrappers.
- add code in catalua.cpp (probably to the existing LuaValue/LuaReference or make a new type or something else) to support this. That's probably the most complicated part.

**Comment by macrosblackd** [comment]
I've commented out the sun* functions in the lua def for now.

**Comment by Malkeus** [comment]
Windows 7, 64bit in 0.C-4244
Just merged the latest changes to this commit, my original STS version survivor managed to gain a skill level without crashing! https://i.imgur.com/lDWIaE2.png
I can't wait to take some of the new lua functions for a spin, also. Thanks, macrosblackd.

**Comment by chaosvolt** [comment]
Swiggity swooty. If all goes as planned, I assume that my latest adventurer shall be my last to exploit the "lower base stats for free points" cheat. o3o

**Comment by Malkeus** [comment]
There's no such thing as cheating when you're playing with yourself. If it feels good, do it. And yes, I mean that any way you want to interpret it.

**Comment by chaosvolt** [comment]
[Immature Humor Intensifies]

**Comment by macrosblackd** [comment]
@Malkeus can you play around with mutations that altar stats to see if there are any  issues in that department? Mainly I'm worried about the ones that straight up SET the stat_max rather than adding/subtracting to it. I may wind up finishing the refactor of stat_max to functions and including a callback to capture the changes.

**Comment by Malkeus** [comment]
That would be mutations from the alpha branch, right? I don't know of any others, could you give me some examples if there are any? I'll be at a PC in a couple hours, I'll check it out then.

**Comment by Malkeus** [comment]
Original Stats: https://i.imgur.com/CWvUmun.png

Gave myself Very Strong, seemed ok, strength was at 12. Until I leveled up mechanics and triggered the script: 
https://i.imgur.com/zwTTTpD.png . It seems I've migrated to version 1 again. Strength remains at 12 though, so that's good. I'm going to alt-f4 out and try just leveling up mechanics without the mutation...

Same result without the mutation, migrating from version 1. Further skill gains continue giving the message: https://i.imgur.com/XiWSkAX.png

Where is player:set_value("StatsThoughSkills", tostring(version)) writing the value to? It doesn't appear anywhere in the .sav for this character.

File: data/mods/StatsThroughSkills/main.lua [hunk 3]
```diff
@@ -1,119 +1,138 @@
 --Increasing stats by skill
 local MOD = {}
 
-mods["StatsThoughSkills"] = MOD
+local version = 2
 
-function MOD.on_day_passed()
+local str_skills = { "carpentry", "mechanics", "swimming", "bashing", "cutting", "melee", "throw" }
```
Comment (Malkeus) [inline data/mods/StatsThroughSkills/main.lua]: Is carpentry a skill? Is that the name in the code for construction?

**Comment by Malkeus** [comment]
Base strength was 10, gave myself prime threshold and prime strength, strength is now 15.
Leveled mechanics to trigger the script and strength held steady at 15. I'm not sure if this is valid representation of actual mechanics, since it's still running the new player script and I get the Migrating from version 1 message. 
Next test will be seeing what happens when I get prime strength after STS has boosted strength over 18....Which I can't seem to achieve. Might be a symptom of using the debug menu to give myself skill levels? 
https://i.imgur.com/PrT93Sg.png
If I'm reading the code right, this should give me 6 points setting me to 16 strength, but it set me to 12 instead. Again, might be because of the migrating to version 1 issue.

On a side note, if I read that right I will need 84 skill points divided among 7 skills in order to get the full 6 point bonus? I don't use skill rust, but it seems like it will be pretty tough for people who do.

**Comment by macrosblackd** [comment]
Balancing the actual numbers can come later, I wanna get it working first =D I'll poke around the code some more to see whats going on with the migration issue.

**Comment by Coolthulhu** [comment]
By the way, Prime [stat] mutations will not work well with this mod.
This can't be fixed from lua alone without representing mutation stats in lua.

**Comment by macrosblackd** [comment]
That's something I've noticed. I'm open to any suggestions on how to handle this. Ideally mutations shouldn't just SET a stat to something, but instead, would adjust it.

**Comment by Malkeus** [comment]
Ah, I'm glad those aren't the final numbers. I know the old numbers could get ridiculous, but it was fun and made grinding skill levels more rewarding.

**Comment by macrosblackd** [comment]
I'm thinking of maybe removing the cap and using a logarithmic function to slow the stat gains at higher skill points.

**Comment by macrosblackd** [comment]
The latest push should prevent the constant migration issue.

**Comment by Malkeus** [comment]
The prime mutations are based on current stats, right? Iirc it was a series of if stat is >= blah then stat == blah, similar to your skill_total settings

**Comment by Malkeus** [comment]
ah, very nice. I stared at main.lua for 20 minutes and never noticed that

**Comment by Malkeus** [comment]
Ok, redoing the tests: Base strength 10, given prime threshold and prime strength -> strength is 15. Leveling rifles and...strength stays at 15: https://i.imgur.com/Zah3Wj0.png
        "values" : {
            "per_bonus" : "1",
            "int_bonus" : "1",
            "StatsThroughSkills" : "2",
            "str_bonus" : "1",
            "dex_bonus" : "1"
        },

**Comment by macrosblackd** [comment]
can you try adding the prime threshold after you get some stat bonus's and see if they stick?

**Comment by Malkeus** [comment]
Base strength 18, bonus gives me 4 on top of that so 22: https://i.imgur.com/jfytPLP.png
Giving prime threshold and prime strength sets strength to 18: https://i.imgur.com/jfytPLP.png
Further skill leveling doesn't reset the strength level to 22: https://i.imgur.com/TbYnMzK.png

**Comment by Malkeus** [comment]
This has been the way prime has always acted. If you managed to get a stat over 18 through granades, and then got prime in that stat, it set it back to 18, making it a nerf for overstrength characters.

**Comment by Malkeus** [comment]
} else if (mut == "STR_ALPHA") {
        ///\EFFECT_STR_MAX determines bonus from STR mutation
        if (str_max <= 6) {
            str_max = 8;
        } else if (str_max <= 7) {
            str_max = 11;
        } else if (str_max <= 14) {
            str_max = 15;
        } else {
            str_max = 18;
        }
it's this damn else statement. would it be so wrong to add a condition where if the stat was over 18, prime had no effect?

**Comment by Malkeus** [comment]
maybe something like 
else if (str_max <= 18) {
str_max = 18;
} else{ 
str_max = str_max;
}
or whatever statement is appropriate to making no change to the existing stat.

**Comment by Malkeus** [comment]
#15019 I proposed making those changes and Kevin Granade explained that prime is SUPPOSED to work like that. I'd say carry on and don't worry about prime being an issue macrosblackd.

It only sets your stats at the moment you gain it, so any future gains would work as expected.

**Pull request merged**